### PR TITLE
Add G4InfoReducer module

### PR DESCRIPTION
**Pull request by Temigo**
This PR is part of an effort from the ICARUS ML group to merge our pipeline back into the official production workflow.

It creates a new module which reads a list of `sim::SimEnergyDeposit` and outputs a list of `sim::SimEnergyDepositLite` (lighter version). It also can voxelize these lite SED using a fhicl parameter to be coarser than the G4 output.

**What needs to happen before this can be merged:**
I have a PR [here](https://github.com/LArSoft/lardataobj/pull/30) which needs to be merged into LArSoft/lardataobj first, since it introduces the new data object `sim::SimEnergyDepositLite` and we need its header file.

### CODE REVIEW

**Comment by Temigo** [comment]
@miquelnebot @SFBayLaser The `lardataobj` PR that was holding up this G4InfoReducer PR has now been merged and `v09_60_00` is the LArSoft version that includes the necessary changes (need header for `SimEnergyDepositLite.h`). I think this PR is ready for review.

**Comment by Temigo** [comment]
trigger build [SBNSoftware/lardataobj#32](https://github.com/LArSoft/lardataobj/pull/32)

**Comment by Temigo** [comment]
@miquelnebot Looking at the CI report [here](https://dbweb8.fnal.gov:8443/LarCI/app/storage/docs/2022/09/28/_report_checkoutLArSoft%23Uc96Xlq.log) it looks like maybe it failed because `lardataobj` needs to be checked out before `lardataalg` ? I am not sure if this is related to my PR or if that's a change on LArSoft side that needs to happen. What do you think?

The command is (you can see `lardataalg` is checked before `lardataobj`)
```
5: DEBUG cmd: 
6:               for LArSoftmodule in larana larcore larcoreobj larcorealg lardata lardataalg lardataobj lareventdisplay larevt larexamples larpandora larpandoracontent larreco larrecodnn larsim larg4 larsoft larsoftobj larwirecell larsimrad webevd larvecutils larfinder larsimdnn ${LArSoftmodules_extra}
7:               do
8:                  date "+%F %T"
9:                  mdir=`echo $LArSoftmodule | tr - _` && (cd $mdir && module_repourl_git $LArSoftmodule)
10:                  this_res=$?
11:                  if [[ "$this_res" != "0" ]]; then break; fi
12:               done
```
and then the crash after checking `lardataalg`
```
43: Module-version: lardataalg 7a760a944357e16659c655190c203bd326a42df5 	https://github.com/LArSoft/lardataalg/commits LARSOFT_SUITE_v09_60_00
44: 2022-09-28 01:53:52
45: /scratch/workspace/lar_ci/label_exp/SLF7/label_exp2/swarm/generic_ci/bin/workflow.sh: line 3486: cd: lardataobj: No such file or directory
```

**Comment by Temigo** [comment]
trigger build LArSoft/lardataobj#32

**Comment by Temigo** [comment]
trigger build

**Comment by Temigo** [comment]
@miquelnebot @SFBayLaser I corrected this PR so it passes the CI checks. Could you please assign a reviewer for me? I cannot do it myself (not a member). Thanks!

**Review by SFBayLaser — APPROVED** [review]
I don't have anything to add past Wes' comments, this looks fine to me.

File: sbncode/LArG4/G4InfoReducer_module.cc [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    double x = sed.X() - std::fmod(sed.X() - fMinX, fVoxelSizeX);
+    double y = sed.Y() - std::fmod(sed.Y() - fMinY, fVoxelSizeY);
+    double z = sed.Z() - std::fmod(sed.Z() - fMinZ, fVoxelSizeZ);
+    // Copy info to SimEnergyDepositLite
+    sim::SimEnergyDepositLite sed_lite(sed.E(), geo::Point_t(x, y, z), sed.T(), sed.OrigTrackID());
+    // Attempt to insert, if already exist we sum up energy deposit
+    auto it = sedlite_v2.find(sed_lite);
+    if (it!=sedlite_v2.end()) {
+      double new_energy = sed_lite.E() + it->E();
+      double new_time = std::min(sed_lite.T(), it->T());
```
[the remark this answers is not in the document]
Comment (Temigo) [reply sbncode/LArG4/G4InfoReducer_module.cc:144]: I don't think there is a strong reason. We use the SimEnergyDeposit time to store the time of first/last step (i.e. first / last energy depositions in voxels) but I cannot think of a reason why the sub-voxel time resolution matters. I used the minimum as a default.

File: sbncode/LArG4/G4InfoReducer_module.cc [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+      if (lhs.X() < rhs.X()) return true;
+      if (lhs.X() > rhs.X()) return false;
+      if (lhs.Y() < rhs.Y()) return true;
+      if (lhs.Y() > rhs.Y()) return false;
+      if (lhs.Z() < rhs.Z()) return true;
+      if (lhs.Z() > rhs.Z()) return false;
+      return lhs.TrackID() < rhs.TrackID();
+    }
+  };
+  std::set<sim::SimEnergyDepositLite, comp> sedlite_v2;
```
[the remark this answers is not in the document]
Comment (Temigo) [reply sbncode/LArG4/G4InfoReducer_module.cc:122]: There is already a comparison operator in SimEnergyDepositLite which mirrors the one of SimEnergyDeposit for backwards compatibility. Compared to the one in G4InfoReducer, it also compares time and energy.

**Pull request merged**